daemon: reinit health monitor on live-restore

corhere opened the pull request:
The container may have been running without health probes for an indeterminate amount of time. The container may have become unhealthy in the interim. We should probe it sooner than in steady-state, while also giving it some leeway to recover from e.g. timed-out connections. This is easy to achieve by probing the container like a freshly-started one. The original author of health-checks came to the same conclusion; the health monitor was reinitialized on live-restored containers [before v17.11.0, when health monitoring of live-restored containers was accidentally broken.](https://github.com/moby/moby/pull/34895/commits/ddae20c032058a0fd42c34c2e9750ee8f6296ac8#diff-9e6623b101327f7efd4cd505efc789f6c0cbd4b150df7af0efe212739d31d861L118-L125) When the regression was noticed (for v23.0.0, _five years_ after it was broken!) [health monitoring of live-restored containers was resumed](https://github.com/moby/moby/pull/41935), not reinitialized. With the new addition of `Healthcheck.StartInterval`, the original behaviour is even more compelling.

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Revert to the original behavior of reinitializing the health monitor on live restore so that restored containers' health monitors are reset to the start period.

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
- Containers that are live-restored will now be given another health-check start period when the daemon restarts, giving the container more time to recover from live-restore before the health-retries countdown begins while being health-checked at the start interval.

**- A picture of a cute animal (not mandatory but encouraged)**

corhere committed:
daemon: reinit health monitor on live-restore

The container may have been running without health probes for an
indeterminate amount of time. The container may have become unhealthy in
the interim. We should probe it sooner than in steady-state, while also
giving it some leeway to recover from e.g. timed-out connections. This
is easy to achieve by probing the container like a freshly-started one.
The original author of health-checks came to the same conclusion; the
health monitor was reinitialized on live-restored containers before
v17.11.0, when health monitoring of live-restored containers was
accidentally broken. Revert to the original behavior.

Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah reviewed (approved):
LGTM

but perhaps @cpuguy83 could have a quick peek to double-check if there's any concerns

cpuguy83 commented:
Do we have a test for this in live-restore?
A quick search doesn't yield anything:

```
$ git grep -i 'Live' | grep -i Health
$ git grep -i 'Restore' | grep -i Health
daemon/health.go:361:// Reset the health state for a newly-started, restarted or restored container.
```

corhere committed:
integration: test container healthcheck is reset

Update the TestDaemonRestartKilContainers integration test to assert
that a container's healthcheck status is always reset to the Starting
state after a daemon restart, even when the container is live-restored.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere commented:
@cpuguy83 an integration test was added as part of #41935. I have now updated that test to also assert that the health monitor is reset to the `Starting` state when the container is (live-)restored after a daemon restart.
